src/mesh/meshblock.cpp

```diff
@@ -164,11 +164,119 @@ MeshBlock::MeshBlock(int igid, int ilid, LogicalLocation iloc, RegionSize input_
+                     MeshBlock *lastBlock)
+    : exec_space(DevExecSpace()), pmy_mesh(pm), loc(iloc), block_size(input_block),
+      gid(igid), lid(ilid), gflag(igflag), properties(properties), packages(packages),
+      prev(lastBlock), next(nullptr), new_block_dt_{}, new_block_dt_hyperbolic_{},
+      new_block_dt_parabolic_{}, new_block_dt_user_{}, cost_(1.0) {
+
+  // set last block's pointer
+  if (lastBlock != nullptr) {
+    lastBlock->next = this;
+  }
+  // initialize grid indices
+  // Allow for user overrides to default Parthenon functions
+  if (app_in->InitApplicationMeshBlockData != nullptr) {
+    InitApplicationMeshBlockData = app_in->InitApplicationMeshBlockData;
+  }
+  if (app_in->InitUserMeshBlockData != nullptr) {
+    InitUserMeshBlockData = app_in->InitUserMeshBlockData;
+  }
+  if (app_in->ProblemGenerator != nullptr) {
+    ProblemGenerator = app_in->ProblemGenerator;
+  }
+  if (app_in->MeshBlockUserWorkInLoop != nullptr) {
+    UserWorkInLoop = app_in->MeshBlockUserWorkInLoop;
+  }
+  if (app_in->UserWorkBeforeOutput != nullptr) {
+    UserWorkBeforeOutput = app_in->UserWorkBeforeOutput;
+  }
+
+  // std::cerr << "WHY AM I HERE???" << std::endl;
+
+  // initialize grid indices
+  if (pmy_mesh->ndim >= 3) {
+    InitializeIndexShapes(block_size.nx1, block_size.nx2, block_size.nx3);
+  } else if (pmy_mesh->ndim >= 2) {
+    InitializeIndexShapes(block_size.nx1, block_size.nx2, 0);
+  } else {
+    InitializeIndexShapes(block_size.nx1, 0, 0);
+  }
+
+  auto &real_container = real_containers.Get();
+  // Set the block pointer for the containers
+  real_container->setBlock(this);
+
+  // (probably don't need to preallocate space for references in these vectors)
+  vars_cc_.reserve(3);
+  vars_fc_.reserve(3);
+
+  // (re-)create mesh-related objects in MeshBlock
+  coords = Coordinates_t(block_size, pin);
+  auto xmin = coords.GetXmin();
+  // std::cout << "XY:" << igid << ":" << iloc.lx1 << ":" << iloc.lx2 << ":" << iloc.level
+  //           << ":" << xmin[0] << ":" << xmin[1] << std::endl;
+
+  // Boundary
+  pbval = std::make_unique<BoundaryValues>(this, input_bcs, pin);
+
+  // Reconstruction (constructor may implicitly depend on Coordinates)
+  precon = std::make_unique<Reconstruction>(this, pin);
+
+  // Add field properties data
+  for (int i = 0; i < properties.size(); i++) {
+    StateDescriptor &state = properties[i]->State();
+    for (auto const &q : state.AllFields()) {
+      real_container->Add(q.first, q.second);
+    }
+    for (auto const &q : state.AllSparseFields()) {
+      for (auto const &m : q.second) {
+        real_container->Add(q.first, m);
+      }
+    }
+  }
+  // Add physics data
+  for (auto const &pkg : packages) {
+    for (auto const &q : pkg.second->AllFields()) {
+      real_container->Add(q.first, q.second);
+    }
+    for (auto const &q : pkg.second->AllSparseFields()) {
+      for (auto const &m : q.second) {
+        real_container->Add(q.first, m);
+      }
+    }
+  }
```

Comment:
My feeling is we repeat a lot of this code in a lot of the meshblock constructors. Maybe we can abstract it away into a helper function? This is non-blocking. Just food for thought.

Reply:
Agreed, but this is one I'm not going to fix in this MR.
It deserves its own MR.

Reply:
Works for me.